contracts/access/manager/utils/AccessManagedAdapter.sol

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.19;
+
+import {AccessManaged} from "../AccessManaged.sol";
+
+/**
+ * @dev This contract can be used to migrate existing {Ownable} or {AccessControl} contracts into an {AccessManager}
+ * system.
+ *
+ * Ownable contracts can have their ownership transferred to an instance of this adapter. AccessControl contracts can
+ * grant all roles to the adapter, while ideally revoking them from all other accounts. Subsequently, the permissions
+ * for those contracts can be managed centrally and with function granularity in the {AccessManager} instance the
+ * adapter is connected to.
+ *
+ * Permissioned interactions with thus migrated contracts must go through the adapter's {relay} function and will
+ * proceed if the function is allowed for the caller in the AccessManager instance.
+ */
+contract AccessManagedAdapter is AccessManaged {
+    /**
+     * @dev Initializes an adapter connected to an AccessManager instance.
+     */
+    constructor(address initialAuthority) AccessManaged(initialAuthority) {}
+
+    /**
+     * @dev Relays a function call to the target contract. The call will be relayed if the AccessManager allows the
+     * caller access to this function in the target contract, i.e. if the caller is a member of the group that is
+     * allowed for the function.
+     */
+    function relay(address target, bytes calldata data) external payable {
```

Comment:
I think we need to warn users to not authorize any group to call `relay`. Imagine the following scenario:

- There's a `ProxyAdmin` owned by the `AccessManagedAdapter`
- There are two participants in the system: Alice and Bob
- Alice can call `upgradeAndCall` in the `ProxyAdmin` but Bob can't
- Additionally, this adapter itself is allowed to call `upgradeAndCall`
- Bob, on the other hand, is allowed to call `relay`

Because of its ability to call `relay`, Bob can reenter this function by calling relay so that `_msgSender()` is the address of the adapter who's also allowed to call `upgradeAndCall`.

Perhaps the scenario is a bit far fetched but I'd expect some cases. We may want to check at least that `_msgSender() != address(this)`

```suggestion
    function relay(address target, bytes calldata data) external payable {
        if(_msgSender() == address(this)) revert AccessManagedAdapterReentrantCall();
```

Reply:
I'm not sure I follow.

- The `ProxyAdmin` is owner by the `AccessManagedAdapter`. So the only way to perform an upgrade is through the adapter's `relay` function.
- When calling the `relay` function, the adapter checks that the `_msgSender()`  is authorize to call the targeted function. In that case that means being allowed to call the `upgradeAndCall` on the `ProxyAdmin`

You are worried that someone may set the system such that:
- the `upgradeAndCall` on the `ProxyAdmin` is set so that you need the `UPGRADE` group to call it.
- the adapter is in the `UPGRADE` group (with Alice)
- the `relay` function on the adapter is set so that you need `SOME_GROUP` to call it (even though the function is not restricted and that is not enforced)
- Bob is added to `SOME_GROUP`

So that bob can call `relay` directly, with `target = adapter` and data that is a `relay` call, that way the adapter would allow the operation, since Bob is authorized to call relay ... and that would reenter, and execute with `msgSender = adapter` ... which is authorized to do an upgrade?

I find that very unlikelly to happen by mistake. You have to screw up twice: 
- give the right to upgrade to the adapter: that error could honestly happen
- setup call permissions for the relay function in the access manager: that is very unlikelly.

If we thing that is a serious issue, then why do 
```
if (_msgSender() == address(this)) revert AccessManagedAdapterReentrantCall();
```
and not
```
if (target == address(this)) revert AccessManagedAdapterReentrantCall();
```
?

@frangio WDYT? Are you concerned by this?

Reply:
`if (target == address(this)) revert AccessManagedAdapterInvalidTarget();` makes sense IMO.

Reply:
`target == address(this)` is possibly something we will need if we want to use the schedule+relay to enforce delays on the admin function

Reply:
Something that we may want to do (I wanted to add that to a followup PR, but if needed lets do it here)

In schedule
```solidity
if (target == address(this)) {
   // check that caller is an admin
   // use custom delay for admin functions
} else {
  // check that caller can call
  // use delay for normal indirect operations
}
// schedule normaly
```

In relay
```solidity
if (target == address(this)) {
   revert SomeError();
}
// relay normaly
```

And add a modifier that replaces `onlyGroup(ADMIN_GROUP)` that enforce that the call:
- comes from an admin
- apply the relay logic.

That way you can only shedule a call to the manager is you are an admin, and you cannot use relay to target the manager.

Would that be good ? Should we do it in this PR?

Reply:
One part that we missed is the current design of the `canCall` function

```soldity
    function canCall(address caller, address target, bytes4 selector) public view virtual returns (bool, uint32) {
        AccessMode mode = getContractMode(target);
        if (mode == AccessMode.Open) {
            return (true, 0);
        } else if (mode == AccessMode.Closed) {
            return (false, 0);
        } else if (caller == address(this)) {
            // Caller is AccessManager => call was relayed. In that case the relay already checked permissions. We
            // verify that the call "identifier", which is set during the relay call, is correct.
            return (_relayIdentifier == keccak256(abi.encodePacked(target, selector)), 0);
        } else {
            // ...
        }
    }
```
if the caller is the manager (the assumption is that is the case only through relay), we don't check any group, we only check that the call is the result of a relay that was registered in the _relayIdentifier.

This means that granting permissions (group) to `address(this)` has absolutelly no effect on the outcome of `canCall`. We can block that in `_grantGroup`, so that there is a verbose error explaining that fact, but as fact as security goes, that is not a concern

Reply:
Note: The above comments are about AccessManager but this thread was about AccessManagerAdapter.

Reply:
es, I got confused in so many convesations :/

Reply:
My point is that we should document better the capabilities of the `relay` function. Because if someone is allowed to use it, they can reenter so that any permissions given to the manager are given to the `relay` caller.

Although I agree it's not likely, this is a valid issue that may come from the audit, I just want to make sure we're already aware of that.

Reply:
This is fixed in the latest code because:
- Not possible to assign a family to AccessManager, thus not possible to assign groups to functions
- Not possible to "close" the AccessManager contract
- Relay and Schedule implement special logic for the AccessManager